beacon_node/network/src/sync/network_context/custody_by_root.rs

```diff
@@ -0,0 +1,417 @@
+            ),
+            active_batch_columns_requests: <_>::default(),
+            failed_peers: LRUTimeCache::new(Duration::from_secs(FAILED_PEERS_CACHE_EXPIRY_SECONDS)),
+            log,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Insert a downloaded column into an active custody request. Then make progress on the
+    /// entire request.
+    ///
+    /// ### Returns
+    ///
+    /// - `Err`: Custody request has failed and will be dropped
+    /// - `Ok(Some)`: Custody request has successfully completed and will be dropped
+    /// - `Ok(None)`: Custody request still active
+    pub(crate) fn on_data_column_downloaded(
+        &mut self,
+        peer_id: PeerId,
+        req_id: DataColumnsByRootRequestId,
+        resp: RpcResponseResult<DataColumnSidecarVec<T::EthSpec>>,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> CustodyRequestResult<T::EthSpec> {
+        // TODO(das): Should downscore peers for verify errors here
+
+        let Some(batch_request) = self.active_batch_columns_requests.get_mut(&req_id) else {
+            warn!(self.log,
+                "Received custody column response for unrequested index";
+                "id" => ?self.requester,
+                "block_root" => ?self.block_root,
+                "req_id" => %req_id,
+            );
+            return Ok(None);
+        };
+
+        match resp {
+            Ok((data_columns, _seen_timestamp)) => {
+                debug!(self.log,
+                    "Custody column download success";
+                    "id" => ?self.requester,
+                    "block_root" => ?self.block_root,
+                    "req_id" => %req_id,
+                    "peer" => %peer_id,
+                    "count" => data_columns.len()
+                );
+
+                // Map columns by index as an optimization to not loop the returned list on each
+                // requested index. The worse case is 128 loops over a 128 item vec + mutation to
+                // drop the consumed columns.
+                let mut data_columns = HashMap::<ColumnIndex, _>::from_iter(
+                    data_columns.into_iter().map(|d| (d.index, d)),
+                );
+                // Accumulate columns that the peer does not have to issue a single log per request
+                let mut missing_column_indexes = vec![];
+
+                for column_index in &batch_request.indices {
+                    let column_request = self
+                        .column_requests
+                        .get_mut(column_index)
+                        .ok_or(Error::BadState("unknown column_index".to_owned()))?;
+
+                    if let Some(data_column) = data_columns.remove(column_index) {
+                        column_request.on_download_success(
+                            req_id,
+                            peer_id,
+                            // Safe to cast, self.column_requests only contains indexes for columns we must custody
```

Comment:
This comment can be removed as we're no longer casting, although I'm not sure why we removed the `CustodyDataColumn` type